make index assembling change

JaredCrean2 opened the pull request:
The changes indexing to *not* call `AssemblyBegin` and `AssemblyEnd` after each `SetValues` call, because this causes deadlocks in parallel.  Now the user can call `SetValues` as many time as needed and call `AssemblyBegin`/`AssemblyEnd` afterwards.  This is also a performance benefit, because `AssemblyBegin` and `AssemblyEnd` do a lot of parallel communication.

stevengj commented:
There should be a custom `show(io::IO, v::Vec)` method etcetera that prints the size and the assembling status if it is still being assembled, and calls `invoke(show, (IO, AbstractVector), io, v)` otherwise.

stevengj commented on src/mat.jl:
```diff
@@ -564,8 +562,8 @@ end
-  AssemblyBegin(x, t)
-  AssemblyEnd(x,t)
+    AssemblyBegin(x, t)
+    AssemblyEnd(x,t)
```
changing the indenting?  

Better to be consistent ... I use an editor (Atom) with a plugin that detects the current indentation setting.

JaredCrean2 commented:
I added a `show` method, but its not being called by whatever magics the REPL uses to create output:

```julia
julia> v = Vec(Float64, 5)
5-element PETSc.Vec{Float64,:mpi}:
 0.0
 0.0
 0.0
 0.0
 0.0

julia> show(STDOUT, v)
Petsc Vec of lenth 5
Process 0 not assembled
```

JaredCrean2 commented:
Also,  I updated the assembly state checking a bit.  For matrices, each Petsc process keeps track of whether its local piece of the matrix is assembled, so I used that instead of tracking it ourselves (this is also more robust, because it still works even if users call one of the variants of `C.*SetValues*`.  I added a field `verify_assembled` to control whether `isassembled` does an Allreduce to check all processes or only checks the local one.

For vectors Petsc does not track the assembly state, so we have to do that ourselves.  Unfortunately, it means users can't call `C.*SetValues*` and have the state checking still work.  Setting `verify_assembled` to false is pretty much a promise by the user to do assembly themselves.

JaredCrean2 commented:
Also, the tests are going to fail because my PR to add `Allreduce` to MPI.jl hasn't been merged yet.  The tests pass locally though.

stevengj commented:
`type Foo; end; Base.show(io::IO, f::Foo) = println(io, "got a Foo"); Foo()` works for me.

stevengj commented:
I think in Julia 0.4 you have to overwrite `writemime(io::IO, ::MIME"text/plain", x::Vec)`.

JaredCrean2 commented:
`writememe` did it.  I build 0.5, and it just overloading `show` was enough.

JaredCrean2 commented:
`allreduce` PR merged in `MPI.jl`, restarting travis.

JaredCrean2 commented:
Linux travis now passes.  Filed an issue for OSX in JuliaParallel/MPI.jl#163

JaredCrean2 commented:
The more I investigate this the weirder it gets.  I determined this is an OpenMPI vs MPICH issue, not OSX vs Linux.  There was a problem with how we were getting the communicator of a Petsc object, and I created a temporary fix, but that didn't solve the problem.  More interestingly, even when using MPICH, the communicator returned by the `comm(vec)` function is not the communicator used to create the vector.  For whatever reason, when using MPICH that communicator works correctly for doing the `allreduce`, but it doesn't for OpenMPI.

JaredCrean2 commented:
I found a minimal reproducible example:

```julia
using MPI

MPI.Init()

val = Int32(false)  # Int8 works
val = MPI.Allreduce(val, MPI.LAND, MPI.COMM_WORLD)
println("val = ", val)
```
produces output:
```
[dev1-VirtualBox:15438] *** An error occurred in MPI_Allreduce: the reduction operation MPI_LAND is not defined on the MPI_INTEGER4 datatype
[dev1-VirtualBox:15438] *** reported by process [1085734913,0]
[dev1-VirtualBox:15438] *** on communicator MPI_COMM_WORLD
[dev1-VirtualBox:15438] *** MPI_ERR_OP: invalid reduce operation
[dev1-VirtualBox:15438] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
```

Weirdly, if we let Petsc initialize MPI instead of doing it ourselves, the error message doesn't appear and the allreduce completes successfully in the minimal example, though not in the original problem. I suspect the minimal example would fail on OSX, but I don't have an OSX system to test on.

Relatedly, in the course of debuggin this little nightmare of a bug, I found some code in `petscsys.h` that can only describe really shady.  It appears Petsc intercepts MPI calls by defining macros with the same name as MPI functions and, in some but not all cases, passes the calls through to the actual MPI functions (why this doesn't cause infinite recursion in the C preprocessor I don't know)

Regardless, the solution is to use Int8s to do the reduction rather than Int32s.
